akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala

```diff
@@ -0,0 +1,287 @@
+/**
+ * Copyright (C) 2016-2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed
+package internal
+
+import java.util.concurrent.ThreadLocalRandom
+
+import akka.actor.DeadLetterSuppression
+import akka.actor.typed.BehaviorInterceptor.SignalTarget
+import akka.actor.typed.SupervisorStrategy._
+import akka.actor.typed.scaladsl.Behaviors
+import akka.annotation.InternalApi
+import akka.util.OptionVal
+
+import scala.concurrent.duration.{ Deadline, FiniteDuration }
+import scala.reflect.ClassTag
+import scala.util.control.Exception.Catcher
+import scala.util.control.NonFatal
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object Supervisor {
+  def apply[T, Thr <: Throwable: ClassTag](initialBehavior: Behavior[T], strategy: SupervisorStrategy): Behavior[T] = {
+    strategy match {
+      case r: Resume ⇒
+        Behaviors.intercept[T, T](new ResumeSupervisor(r))(initialBehavior)
+      case r: Restart ⇒
+        Behaviors.intercept[T, T](new RestartSupervisor(initialBehavior, r))(initialBehavior)
+      case r: Stop ⇒
+        Behaviors.intercept[T, T](new StopSupervisor(initialBehavior, r))(initialBehavior)
+      case r: Backoff ⇒
+        Behaviors.intercept[AnyRef, T](new BackoffSupervisor(initialBehavior, r))(initialBehavior).asInstanceOf[Behavior[T]]
+    }
+  }
+}
+
+abstract class AbstractSupervisor[O, I, Thr <: Throwable](ss: SupervisorStrategy)(implicit ev: ClassTag[Thr]) extends BehaviorInterceptor[O, I] {
```

Comment:
`@InternalApi private[akka]`, for all these classes

`strategy` would be nicer than `ss`